core/src/com/unciv/logic/map/MapGenerator.kt

```diff
@@ -371,15 +372,66 @@ class MapGenerator(val ruleset: Ruleset) {
+        val allBaseOccursOn = allOccursOn.map { ruleset.terrains[it]!! }.filter { it.type == TerrainType.Water || it.type == TerrainType.Land }.map { it.name }.toHashSet()
+        val allFeaturesOccursOn = allOccursOn.map { ruleset.terrains[it]!! }.filter { it.type == TerrainType.TerrainFeature }.map { it.name }.toHashSet()
+
+        // Get filtered sequences for allowable tiles without resources yet (nat wonders aren't spawned yet) - for all domain types
+        // Could be simplified by basing landCandidates/waterCandidates on allCandidates, but it should enumerate faster with cheapest tests first
+        val allCandidates =
+                tileMap.values.asSequence().filter {
+                    it.resource == null
+                    && (it.baseTerrain in allBaseOccursOn || it.terrainFeature in allFeaturesOccursOn) }
+        val landCandidates =
+                tileMap.values.asSequence().filter {
+                    it.isLand && it.resource == null
+                    && (it.baseTerrain in allBaseOccursOn || it.terrainFeature in allFeaturesOccursOn) }
+        val waterCandidates =
+                tileMap.values.asSequence().filter {
+                    it.isWater && it.resource == null
+                    && (it.baseTerrain in allBaseOccursOn || it.terrainFeature in allFeaturesOccursOn) }
+
+        // This is a good formula taking distributions over land/water into account and measuring precisely, but at the cost of enumerating the candidates one extra time
+        // measurement says it's fast so i'll not replace it with a guesstimate.
+        val totalNumberOfResources = (
+                landCandidates.count() * landOrWaterMap.count { it.value.land }
+                + waterCandidates.count() * landOrWaterMap.count { it.value.water }
+            ) * tileMap.mapParameters.resourceRichness / landOrWaterMap.size
+        val resourcesPerType = (totalNumberOfResources / strategicResources.size).toInt()
+
+        // finally, place the resources type by type - below is the first time larger lists of tiles are actually materialized as chooseSpreadOutLocations needs them
         for (resource in strategicResources) {
-            // remove the tiles where previous resources have been placed
-            val suitableTiles = candidateTiles
-                    .filter { it.resource == null && resource.terrainsCanBeFoundOn.contains(it.getBaseTerrain().name)}
+            // Filter land, sea or both candidates for suitable tiles without those where previous resources have been placed
+            // as the candidates are sequences, they start over every time, therefore taking resources already placed into account - no need to repeat the it.resource == null test
+            val suitableTiles =
+                    (if(landOrWaterMap[resource.name]?.land == false) waterCandidates else if(landOrWaterMap[resource.name]?.water == false) landCandidates else allCandidates)
```

Comment:
This line is unreadable, this should probably be split to a separate val

Reply:
Good point